skunkworks_crow/src/main/java/org/odk/share/tasks/DownloadJob.java

```diff
@@ -121,27 +121,29 @@ private void initJob(Params params) {
                 break;
         }
 
-        setupDataStreamsAndReceive(method, targetMacAddress);
+        setupDataStreamsAndReceive(method);
     }
 
-    private void setupDataStreamsAndReceive(@Share.TransferMethod int method, String macAddress) {
+    private void setupDataStreamsAndReceive(@Share.TransferMethod int method) {
         try {
             Timber.d("Waiting for sender");
-            if (method == Share.TransferMethod.BLUETOOTH && macAddress != null) {
+            if (method == Share.TransferMethod.BLUETOOTH && targetMacAddress != null) {
                 BluetoothAdapter bluetoothAdapter = BluetoothAdapter.getDefaultAdapter();
-                BluetoothDevice bluetoothDevice = bluetoothAdapter.getRemoteDevice(macAddress);
-                bluetoothSocket = bluetoothDevice.createRfcommSocketToServiceRecord(SPP_UUID);
+                BluetoothDevice bluetoothDevice = bluetoothAdapter.getRemoteDevice(targetMacAddress);
+                if (bluetoothDevice != null) {
+                    bluetoothSocket = bluetoothDevice.createRfcommSocketToServiceRecord(SPP_UUID);
 
-                if (!bluetoothSocket.isConnected()) {
-                    bluetoothSocket.connect();
-                }
+                    if (!bluetoothSocket.isConnected()) {
+                        bluetoothSocket.connect();
+                    }
 
-                if (bluetoothSocket.isConnected()) {
-                    rxEventBus.post(new BluetoothEvent(BluetoothEvent.Status.CONNECTED));
-                }
+                    if (bluetoothSocket.isConnected()) {
```

Comment:
this can be replaced with `else`

Reply:
I don't this should go in else because previous if block says it to connect so I think this `if` stmt check can be removed and we can directly write 
` rxEventBus.post(new BluetoothEvent(BluetoothEvent.Status.CONNECTED));`
as it is expected that
```
if (!bluetoothSocket.isConnected()) {
    bluetoothSocket.connect();
}
```
will ensure that bluetooth is connected and if by any chance it fails then it should throw some error or callback that we need to handle explicitly.

Reply:
agreed with @lakshyagupta21 , just removing the `if-else`.

Reply:
Is `bluetoothSocket.connect();` a synchronous event? Are there any callbacks for bluetooth connected or disconnected? If not then we need to fix it.

Reply:
Second problem is that we shouldn't establish bluetooth socket connection within `DownloadJob`. DownloadJob's purpose is to start download task and it shouldn't be doing anything else. This would testing the code in the later part difficult.

Reply:
> Is `bluetoothSocket.connect();` a synchronous event? Are there any callbacks for bluetooth connected or disconnected? If not then we need to fix it.

This method is synchronous yes, and if the connection cannot be established, it will throw an IO Exception. So we don't have to fix that.

Reply:
> Second problem is that we shouldn't establish bluetooth socket connection within `DownloadJob`. DownloadJob's purpose is to start download task and it shouldn't be doing anything else. This would testing the code in the later part difficult.

An issue is, we cannot pass an object as a parameter into our jobs, I tried to establish the socket connection in the activity but it seems the job is not relay on that. Besides, the socket and stream for hotspot are created in the job as well, so I think it doesn't matter.

Reply:
@shobhitagarwal1612 Can you suggest a forward, I tried the workarounds but I'm not able to think one because DownloadJob needs socket objects and we cannot pass the socket objects directly in the `DownloadJob` and `UploadJob`.

Reply:
Okay. I'll give it a try and post it here.